File: medium/53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = function(nums) {
    let maxSum = nums[0];
    let curSum = 0;
    nums.forEach(n => {
        curSum = Math.max(0, curSum);
        curSum += n;
        maxSum = Math.max(curSum, maxSum);
    })
    return maxSum;
};

File: medium/53-maximum-subarray/53-maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
    max_sum = nums[0]
    cur_sum = 0
    nums.each do |n|
        cur_sum = [0, cur_sum].max
        cur_sum += n
        max_sum = [max_sum, cur_sum].max
    end
    max_sum
end